pkg/diagnostics/error.go

```diff
@@ -0,0 +1,80 @@
+package diagnostics
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Err wraps diagnostics information for an error.
+// Err allows providing information like source, reason and message
+// that provides a much better user error reporting capability.
+type Err struct {
+	Orig error
+
+	// Source defines with entity is generating the error.
+	// It allows passing along information about where the error is being
+	// generated from. for example, the Asset.
+	Source string
+
+	// Reason is a CamelCase string that summarizes the error in one word.
+	// It allows easy catgeorizations of known errors.
+	Reason string
+
+	// Message is free-form strings which provides important details or
+	// diagnostics for the error. When writing messages, make sure to keep in mind
+	// that the audience for message is end-users who might not be experts.
+	Message string
+}
+
+// Unwrap allows the error to be unwrapped.
+func (e *Err) Unwrap() error { return e.Orig }
+
+// Error returns a string representation of the Err. The returned value
+// is expected to be a single value.
+// The format of the error string returned is,
+// `error(<Reason>) from <Source>: <Message>: <Cause of Orig>`
+func (e *Err) Error() string {
+	buf := &bytes.Buffer{}
+	if len(e.Source) > 0 {
+		fmt.Fprintf(buf, "error(%s) from %q", e.Reason, e.Source)
```

Comment:
Do we need `%q` for `Source`, or can we use `%s` (here and in your later `Error from %q`)?  Both of those strings are compiled into the installer right?  I'd expect to use %q only if we were passing along some information slurped from the provider, where we weren't sure if there was going to be whitespace or other potentially confusing characters.

Reply:
fixed.